backends/mixer/atari/atari-mixer.cpp

```diff
@@ -0,0 +1,227 @@
+		_clk = CLK10K;
+	} else {
+		_outputRate = 8195;
+		_clk = CLK8K;
+	}
+
+	ConfMan.setInt("output_rate", _outputRate);
+	debug("setting %d Hz mixing frequency", _outputRate);
+
+	_samples = 8192;
+	while (_samples * 16 > _outputRate * 2)
+		_samples >>= 1;
+
+	ConfMan.registerDefault("output_samples", (int)_samples);
+
+	int samples = ConfMan.getInt("output_samples");
+	if (samples > 0)
+		_samples = samples;
+
+	ConfMan.setInt("output_samples", (int)_samples);
+	debug("sample buffer size: %d", _samples);
+
+	_samplesBuf = new uint8[_samples * 4];
+
+	g_system->getEventManager()->getEventDispatcher()->registerObserver(this, 10, false);
+}
+
+AtariMixerManager::~AtariMixerManager() {
+	g_system->getEventManager()->getEventDispatcher()->unregisterObserver(this);
+
+	Buffoper(0x00);
+
+	Mfree(_atariSampleBuffer);
+	_atariSampleBuffer = _atariPhysicalSampleBuffer = _atariLogicalSampleBuffer = nullptr;
+
+	Sndstatus(SND_RESET);
+
+	Unlocksnd();
+
+	_atariInitialized = false;
+
+	delete[] _samplesBuf;
+}
+
+void AtariMixerManager::init() {
+	_mixer = new Audio::MixerImpl(_outputRate, _samples);
+	_mixer->setReady(true);
+
+	_atariSampleBufferSize = _samples * 4;
+
+	_atariSampleBuffer = (byte*)Mxalloc(_atariSampleBufferSize * 2, MX_STRAM);
+	if (!_atariSampleBuffer)
+		return;
+
+	_atariPhysicalSampleBuffer = _atariSampleBuffer;
+	_atariLogicalSampleBuffer = _atariSampleBuffer + _atariSampleBufferSize;
+
+	memset(_atariSampleBuffer, 0, 2 * _atariSampleBufferSize);
+
+	if (Locksnd() < 0)
+		return;
+
+	Sndstatus(SND_RESET);
+	Setmode(MODE_STEREO16);
+	Devconnect(DMAPLAY, DAC, CLK25M, _clk, NO_SHAKE);
+	Soundcmd(ADDERIN, MATIN);
+	Setbuffer(SR_PLAY, _atariSampleBuffer, _atariSampleBuffer + 2 * _atariSampleBufferSize);
+	Buffoper(SB_PLA_ENA | SB_PLA_RPT);
+
+	_atariInitialized = true;
+}
+
+void AtariMixerManager::suspendAudio() {
+	{Common::String str = Common::String::format("suspendAudio\n");
+	g_system->logMessage(LogMessageType::kDebug, str.c_str());}
+
+	Buffoper(0x00);
+
+	_audioSuspended = true;
+}
+
+int AtariMixerManager::resumeAudio() {
+	{Common::String str = Common::String::format("resumeAudio 1\n");
+	g_system->logMessage(LogMessageType::kDebug, str.c_str());}
+
+	if (!_audioSuspended || !_atariInitialized) {
+		return -2;
+	}
+
+	{Common::String str = Common::String::format("resumeAudio 2\n");
+	g_system->logMessage(LogMessageType::kDebug, str.c_str());}
+
+	Buffoper(SB_PLA_ENA | SB_PLA_RPT);
+
+	_audioSuspended = false;
+	return 0;
+}
+
+bool AtariMixerManager::notifyEvent(const Common::Event &event) {
+	switch (event.type) {
+	case Common::EVENT_QUIT:
+	case Common::EVENT_RETURN_TO_LAUNCHER:
+		debug("silencing the mixer");
+		memset(_atariSampleBuffer, 0, 2 * _atariSampleBufferSize);
+		return false;
+	case Common::EVENT_MUTE:
+		_muted = !_muted;
+		debug("audio %s", _muted ? "off" : "on");
+		return false;
+	default:
+		[[fallthrough]];
+	}
+
+	return false;
+}
+
+void AtariMixerManager::update() {
+	if (_audioSuspended && !_muted) {
```

Comment:
Hmm. I agree it's not ideal but it is solely for performance reasons, see: https://github.com/scummvm/scummvm/blob/03438f4e6bfde1718ba577303a61fb6059742985/backends/platform/atari/readme.txt#L315-L326

Basically, as far as I can see, there is no way how to actually force the mixer to be idle, i.e. not eating precious CPU time. For MIDI music you can use the null audio plugin ("No music") but what about CD soundtrack and game sounds? Even if I remove all the *.wav files, it would still mix (muted...) game sounds and therefore wasting cycles.

Reply:
Well, the problem is that those sounds will never be killed, and eventually, free slots will be depleted. More testing is required to make sure that the sound-heavy titles such as SCUMM v7+ will not crash after several minutes of gameplay if set to mute.

Also, I have a question about the platform: is it multitasking? I mean, if it is single-tasking perhaps it will not bring additional benefit to save a few of those CPU cycles.

Reply:
I need to take a look at this later. I agree, the risk you've mentioned is definitely worrying. Yes, the platform is single-tasking. The idea behind this change is that the mix function isn't exactly cheap so why waste it on silence only? It wont create speedup of, say, 10 FPS but even 1-2 FPS can make a difference on a slow platform. :) (I plan to do precise measuring later).

Reply:
I've opened an issue with some general ideas on how to optimise the software mixer, so we can revisit this later.

https://bugs.scummvm.org/ticket/14267